stable_baselines3/common/buffers.py

```diff
@@ -381,3 +381,100 @@ def _get_samples(self, batch_inds: np.ndarray, env: Optional[VecNormalize] = Non
             self.returns[batch_inds].flatten(),
         )
         return RolloutBufferSamples(*tuple(map(self.to_torch, data)))
+
+
+class NstepReplayBuffer(ReplayBuffer):
+    """
+    Replay Buffer that computes N-step returns.
+
+    :param buffer_size: (int) Max number of element in the buffer
+    :param observation_space: (spaces.Space) Observation space
+    :param action_space: (spaces.Space) Action space
+    :param device: (Union[th.device, str]) PyTorch device
+        to which the values will be converted
+    :param n_envs: (int) Number of parallel environments
+    :param optimize_memory_usage: (bool) Enable a memory efficient variant
+        of the replay buffer which reduces by almost a factor two the memory used,
+        at a cost of more complexity.
+        See https://github.com/DLR-RM/stable-baselines3/issues/37#issuecomment-637501195
+        and https://github.com/DLR-RM/stable-baselines3/pull/28#issuecomment-637559274
+    :param n_step: (int) The number of transitions to consider when computing n-step returns
+    :param gamma:  (float) The discount factor for future rewards.
+    """
+
+    def __init__(
+        self,
+        buffer_size: int,
+        observation_space: spaces.Space,
+        action_space: spaces.Space,
+        device: Union[th.device, str] = "cpu",
+        n_envs: int = 1,
+        optimize_memory_usage: bool = False,
+        n_step: int = 1,
+        gamma: float = 0.99,
+    ):
+        super().__init__(buffer_size, observation_space, action_space, device, n_envs, optimize_memory_usage)
+        self.n_step = int(n_step)
+        if not 0 < n_step <= buffer_size:
+            raise ValueError("n_step needs to be strictly smaller than buffer_size, and strictly larger than 0")
+        self.gamma = gamma
+
+    def _get_samples(self, batch_inds: np.ndarray, env: Optional[VecNormalize] = None) -> ReplayBufferSamples:
+        assert not np.any(batch_inds == self.pos)
```

Comment:
why?
Only in the optimize memory case, no?

Reply:
I wanted to make this function accessible without using `self.sample()`.

In `self.sample`, the following prevents `self.pos` from being accessible:

```Python
  if self.full:
            batch_inds = (np.random.randint(1, self.buffer_size, size=batch_size) + self.pos) % self.buffer_size
        else:
            batch_inds = np.random.randint(0, self.pos, size=batch_size)
```

This was done to ensure we can use the same sampling mechanism for both optimized and unoptimized buffers.

Reply:
Hmm,
I had to comment that line because it was giving me an error at some point in the training.

>I wanted to make this function accessible without using self.sample().

Why?

Reply:
It seemed right at the time, but obviously it isn't necessary.